prusti-specs/src/extern_spec_rewriter.rs

```diff
@@ -201,76 +136,470 @@ fn rewrite_method_inputs(item_ty: &syn::Type, method: &mut ImplItemMethod) ->
+                                        .to_token_stream(),
+                                );
+                                continue;
+                            }
+                        }
+                        new_tokens.extend(token.to_token_stream());
+                    }
+                    let res: Result<syn::Item, _> = syn::parse2(new_tokens);
+                    if res.is_err() {
+                        return Err(syn::Error::new(item.span(), "invalid function signature"));
+                    }
+
+                    let mut item = res.unwrap();
+                    if let syn::Item::Fn(item_fn) = &mut item {
+                        rewrite_fn(item_fn, path);
+                    }
+                    *tokens = quote!(#item)
+                }
+                syn::Item::Use(_) => {}
+                _ => return Err(syn::Error::new(item.span(), "unexpected item")),
             }
-            syn::GenericParam::Lifetime(ld) => {
-                let ident = ld.lifetime.clone();
-                syn::parse_quote!{ &#ident ::core::marker::PhantomData<()> }
+        }
+        Ok(())
+    }
+
+    /// Rewrite a specification function to a call to the specified function.
+    /// The result of this rewriting is then parsed in `ExternSpecResolver`.
+    fn rewrite_fn(item_fn: &mut syn::ItemFn, path: &mut syn::Path) {
+        let ident = &item_fn.sig.ident;
+        let args = &item_fn.sig.inputs;
+        let item_fn_span = item_fn.span();
+        item_fn.block = parse_quote_spanned! {item_fn_span=>
+            {
+                #path :: #ident (#args);
+                unimplemented!()
             }
-            syn::GenericParam::Const(_cp) => continue,
         };
-        // Unwrap the `ParsableUnnamedField` and push. See below.
-        fields.push(field.0);
-    }
 
-    new_struct.fields = syn::Fields::Unnamed( syn::parse_quote!{ ( #(#fields),* ) } );
-    Ok(new_struct)
+        item_fn
+            .attrs
+            .push(parse_quote_spanned!(item_fn_span=> #[prusti::extern_spec]));
+        item_fn
+            .attrs
+            .push(parse_quote_spanned!(item_fn_span=> #[trusted]));
+    }
 }
 
-/// The `syn::Field` doesn't implement `syn::parse::Parse` directly since it can do
-/// both `parse_unnamed` and `parse_named`. This is a wrapper to tell `parse_quote`
-/// to use `parse_unnamed` when parsing. https://github.com/dtolnay/syn/issues/651
-struct ParsableUnnamedField(syn::Field);
-impl syn::parse::Parse for ParsableUnnamedField {
-    fn parse(input: syn::parse::ParseStream<'_>) -> syn::parse::Result<Self> {
-        Ok(ParsableUnnamedField( syn::Field::parse_unnamed(input)? ))
+pub mod impls {
+    use super::*;
+
+    pub fn generate_new_struct(item_struct: &syn::ItemImpl) -> syn::Result<syn::ItemStruct> {
+        let name_generator = NameGenerator::new();
+        let struct_name = match name_generator.generate_struct_name(item_struct) {
```

Comment:
Maybe not in scope for this PR, but `NameGenerator::generate_struct_name` should not be returning a `Result`. The [implementation](https://github.com/viperproject/prusti-dev/blob/f7f3e1b27e678c28c078c1be0796ab974dfc7c62/prusti-specs/src/specifications/common.rs#L142-L144) should handle any kind of `syn::Type` and stringify it (somehow) instead of errorring out. This might be one of the reasons why extern specs on slices don't work @JonasAlaif ?